src/include/migraphx/simple_par_for.hpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_RTGLIB_SIMPLE_PAR_FOR_HPP
+#define MIGRAPHX_GUARD_RTGLIB_SIMPLE_PAR_FOR_HPP
+
+#include <thread>
+#include <cmath>
+#include <algorithm>
+#include <vector>
+#include <cassert>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct joinable_thread : std::thread
+{
+    template <class... Xs>
+    joinable_thread(Xs&&... xs) : std::thread(std::forward<Xs>(xs)...) // NOLINT
+    {
+    }
+
+    joinable_thread& operator=(joinable_thread&& other) = default;
+    joinable_thread(joinable_thread&& other)            = default;
+
+    ~joinable_thread()
+    {
+        if(this->joinable())
+            this->join();
+    }
+};
+
+template <class F>
+auto thread_invoke(std::size_t i, std::size_t tid, F f) -> decltype(f(i, tid))
+{
```

Comment:
Is `i` some kind of an index? For laypersons, could you please rename the variable, or document it.. Thanks.

Reply:
`i` is the loop variable or index. Usually this is named `i`.

Reply:
Also this is just moved from `par_for.hpp` so I am not adding this as new.

Reply:
Here it is not a loop variable, but a parameter to thread_invoke, so I would like to know what `i` is, without having to go across functions or files reading code -- and I still don't know what it is here :-)  
Thanks.

Reply:
> Here it is not a loop variable

Its the loop variable that gets passed to this function:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/par_for.hpp#L71

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/par_for.hpp#L90

The variables are named the same as the variables that are passed to make it easier to follow where the variables go in the function. 


> without having to go across functions or files reading code -- and I still don't know what it is here

Thats because you are trying to read the code bottom-up. If you read it top-down then its easier to see how everything is being called.

Reply:
Not to belabor too much.. but I was making a simple point here. That I as a reviewer should be able to look at the changes and make sense of what is going on.. without going too far reading the other code. 

Function parameter names should be named for clarity, and not along how (internal) variables are named/used somewhere else -- that is unintuitive for the uninitiated.. 

We have spent here far more energy and lines of explanation on this simple function parameter `i`. A two line comment there in the code (or the renaming of parameter `i`) would help. But I am not going to discuss this any further.. Thanks.

Reply:
> That I as a reviewer should be able to look at the changes 

But I am not making changes here. The changes is renaming the function from `par_of` to `simple_par_for`. 

> without going too far reading the other code.

You need to read the code to figure out whats going on. This is a fairly simple piece of code(ie invoke a function) that it should be easy to follow.